lib/model/message_list.dart

```diff
@@ -424,13 +440,15 @@ class MessageListView with ChangeNotifier, _MessageSequence {
     _updateEndMarkers();
     notifyListeners();
     try {
+      final generation = this.generation;
       final result = await getMessages(store.connection,
         narrow: narrow.apiEncode(),
         anchor: NumericAnchor(messages[0].id),
         includeAnchor: false,
         numBefore: kMessageListFetchBatchSize,
         numAfter: 0,
       );
+      if (this.generation > generation) return;
```

Comment:
Not sure if we want to skip the finally block as well. At least we shouldn't need to notify the listeners again.
```
    } finally {
      _fetchingOlder = false;
      _fetchingOlder = false;
      _updateEndMarkers();
      _updateEndMarkers();
      notifyListeners();
      notifyListeners();
    }
```

Reply:
In the current revision:
```diff
     } finally {
       _fetchingOlder = false;
-      _updateEndMarkers();
-      notifyListeners();
+      if (this.generation == generation) {
+        _updateEndMarkers();
+        notifyListeners();
+      }
     }
```

The thing is that the reason there's a `notifyListeners` there, and an `_updateEndMarkers`, is as a followup to the `_fetchingOlder` update. Changing `_fetchingOlder` changes what set of end markers should be present; and when the end markers change, listeners should be notified (so that e.g. the UI can update to reflect the change).

We could get an effect like this, because `_fetchingOlder` may already be false. But the way to do that would be with an explicit check on `_fetchingOlder`: no _updateEndMarkers etc. if we're not actually changing the value.

In fact, this actually points to a bug (that was introduced in my draft): if we abort because the generation advanced, then we should *not* set `_fetchingOlder` to false. It might happen to be true by the time we reach this point, because the user might have scrolled up again after the new `fetchInitial` completed, causing a new `fetchOlder`. If it is true after the generation advanced, then the fact that it's true doesn't belong to this `fetchOlder` call, but instead to that newer `fetchOlder` call, and so this `fetchOlder` call shouldn't be interfering.

So please fix that bug from my draft :slightly_smiling_face:, and also add a test that would reproduce the bug.

Reply:
Thanks for the elaboration. I think encapsulating `_fetchingOlder` update with the generation check should be sufficient as a fix.